### PR TITLE
SONARJS-833 Rule: Arguments to built-in functions should match documented types

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArgumentTypesCheck.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    return null;
+  }
+
+  @Override
+  public void endOfExecution(Scope functionScope) {
+    for (Entry<Tree, Optional<String>> entry : hasIssue.entrySet()) {
+      if (entry.getValue().isPresent()) {
+        addIssue(entry.getKey(), entry.getValue().get());
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArgumentTypesCheck.java]: MAJOR Call "isPresent()" before accessing the value. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3655)

File: javascript-frontend/src/main/java/org/sonar/javascript/se/Type.java [hunk 2]
```diff
@@ -55,6 +57,25 @@
   UNDEFINED(Constraint.UNDEFINED, null, new NullOrUndefinedBuiltInProperties()),
   ;
 
+  public static final List<Constraint> ONE_NUMBER = ImmutableList.of(Constraint.ANY_NUMBER);
+  public static final List<Constraint> STRING_NUMBER = ImmutableList.of(Constraint.ANY_STRING, Constraint.ANY_NUMBER);
+  public static final List<Constraint> NUMBER_STRING = ImmutableList.of(Constraint.ANY_NUMBER, Constraint.ANY_STRING);
+  public static final List<Constraint> NUMBER_NUMBER = ImmutableList.of(Constraint.ANY_NUMBER, Constraint.ANY_NUMBER);
+  public static final List<Constraint> EMPTY = ImmutableList.of();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/Type.java]: MINOR Make this member "protected". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2386)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/StringBuiltInProperties.java [hunk 3]
```diff
@@ -19,65 +19,99 @@
...
-      .put("sup", method(Constraint.STRING_PRIMITIVE))
+      .put("anchor", method(Constraint.STRING_PRIMITIVE, ONE_STRING))
+      .put("big", method(Constraint.STRING_PRIMITIVE, EMPTY))
+      .put("blink", method(Constraint.STRING_PRIMITIVE, EMPTY))
+      .put("bold", method(Constraint.STRING_PRIMITIVE, EMPTY))
+      .put("fixed", method(Constraint.STRING_PRIMITIVE, EMPTY))
+      .put("fontcolor", method(Constraint.STRING_PRIMITIVE, ONE_STRING))
+      .put("fontsize", method(Constraint.STRING_PRIMITIVE, ONE_NUMBER))
+      .put("italics", method(Constraint.STRING_PRIMITIVE, EMPTY))
+      .put("small", method(Constraint.STRING_PRIMITIVE, EMPTY))
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/StringBuiltInProperties.java]: @vilchik-elena It seems we don't have [`link`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/link).

File: javascript-frontend/src/main/java/org/sonar/javascript/se/sv/BuiltInFunctionSymbolicValue.java [hunk 4]
```diff
@@ -32,15 +34,24 @@
 
   private final Constraint returnedValueConstraint;
   private final ArgumentsConstrainer argumentsConstrainer;
+  private final IntFunction<Constraint> signature;
+
+  public BuiltInFunctionSymbolicValue(Constraint returnedValueConstraint, IntFunction<Constraint> signature) {
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/sv/BuiltInFunctionSymbolicValue.java]: @vilchik-elena Why not make this constructor call the third one?

Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/sv/BuiltInFunctionSymbolicValue.java]: fixed

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/ArrayBuiltInProperties.java [hunk 5]
```diff
@@ -19,53 +19,70 @@
...
-      .put("pop", method(ANY_VALUE))
-      .put("shift", method(ANY_VALUE))
-      .put("find", method(ANY_VALUE))
-      .put("reduce", method(ANY_VALUE))
-      .put("reduceRight", method(ANY_VALUE))
+      .put("pop", method(ANY_VALUE, Type.EMPTY))
+      .put("shift", method(ANY_VALUE, Type.EMPTY))
+      .put("find", method(ANY_VALUE, functionObjectParameters))
+      .put("reduce", method(ANY_VALUE, ImmutableList.of(FUNCTION, Constraint.ANY_VALUE)))
+      .put("reduceRight", method(ANY_VALUE, ImmutableList.of(FUNCTION, Constraint.ANY_NUMBER)))
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/ArrayBuiltInProperties.java]: fixed

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/StringBuiltInProperties.java (hunk 3 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/StringBuiltInProperties.java]: fixed

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/DateBuiltInProperties.java [hunk 6]
```diff
@@ -96,9 +105,14 @@
   @Override
   Map<String, SymbolicValue> getOwnMethods() {
     return ImmutableMap.<String, SymbolicValue>builder()
-      .put("now", method(Constraint.NUMBER_PRIMITIVE))
-      .put("parse", method(Constraint.NUMBER_PRIMITIVE))
-      .put("UTC", method(Constraint.NUMBER_PRIMITIVE))
+      .put("now", method(Constraint.NUMBER_PRIMITIVE, Type.EMPTY))
+      .put("parse", method(Constraint.NUMBER_PRIMITIVE, Type.ONE_STRING))
+      .put("UTC", method(Constraint.NUMBER_PRIMITIVE, (int index) -> {
+        if (index < 6) {
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/DateBuiltInProperties.java]: fixed

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/ArrayBuiltInProperties.java (hunk 5 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/ArrayBuiltInProperties.java]: fixed

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/ArrayBuiltInProperties.java [hunk 7]
```diff
@@ -19,53 +19,70 @@
...
+import static org.sonar.javascript.se.Constraint.FUNCTION;
+import static org.sonar.javascript.se.Constraint.OBJECT;
 import static org.sonar.javascript.se.Constraint.UNDEFINED;
 
 public class ArrayBuiltInProperties extends BuiltInProperties {
 
   @Override
   Map<String, SymbolicValue> getMethods() {
+    List<Constraint> functionObjectParameters = ImmutableList.of(FUNCTION, OBJECT);
+    List<Constraint> anyValueNumber = ImmutableList.of(ANY_VALUE, ANY_NUMBER);
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/ArrayBuiltInProperties.java]: done

**Pull request merged**